mesonbuild/cargo/interpreter.py

```diff
@@ -0,0 +1,446 @@
+            raw = load_toml(filename)
+
+            raw_manifest = T.cast('manifest.Manifest', raw)
+            man = _convert_manifest(raw_manifest, subdir, m)
+            manifests[man.package.name] = man
+
+    return manifests
+
+
+def load_all_manifests(subproject_dir: str) -> T.Dict[str, Manifest]:
+    """Find all cargo subprojects, and load them
+
+    :param subproject_dir: Directory to look for subprojects in
+    :return: A dictionary of rust project names to Manifests
+    """
+    manifests: T.Dict[str, Manifest] = {}
+    for p in Path(subproject_dir).iterdir():
+        if p.is_dir() and (p / 'Cargo.toml').exists():
+            manifests.update(_load_manifests(str(p)))
+    return manifests
+
+
+def _create_lib(cargo: Manifest, build: builder.Builder) -> T.List[mparser.BaseNode]:
+    kw: T.Dict[str, mparser.BaseNode] = {}
+    if cargo.dependencies:
+        ids = [build.identifier(f'dep_{n}') for n in cargo.dependencies]
+        kw['dependencies'] = build.array(
+            [build.method('get_variable', i, [build.string('dep')]) for i in ids])
+
+    # FIXME: currently assuming that an rlib is being generated, which is
+    # the most common.
+    return [
+        build.assign(
+            build.function(
+                'static_library',
+                [
+                    build.string(fixup_meson_varname(cargo.package.name)),
+                    build.string(os.path.join('src', 'lib.rs')),
+                ],
+                kw,
+            ),
+            'lib'
+        ),
+
+        build.assign(
+            build.function(
+                'declare_dependency',
+                kw={'link_with': build.identifier('lib')} | kw,
+            ),
+            'dep'
+        )
+    ]
+
+
+def interpret(cargo: Manifest, env: Environment) -> mparser.CodeBlockNode:
+    filename = os.path.join(cargo.subdir, cargo.path, 'Cargo.toml')
+    build = builder.Builder(filename)
+
+    ast: T.List[mparser.BaseNode] = [
+        _create_project(cargo.package, build, env),
+        build.assign(build.function('import', [build.string('unstable-rust')]), 'rust'),
+    ]
+
+    if cargo.dependencies:
+        for name, dep in cargo.dependencies.items():
+            kw = {
+                'version': build.array([build.string(s) for s in dep.version]),
+            }
+            ast.extend([
+                build.assign(
+                    build.method(
+                        'cargo',
+                        build.identifier('rust'),
+                        [build.string(name)],
+                        kw,
+                    ),
+                    f'dep_{fixup_meson_varname(name)}',
+                ),
+            ])
+
+    # Libs are always auto-discovered and there's no other way to handle them,
+    # which is unfortunate for reproducability
+    if os.path.exists(os.path.join(env.source_dir, cargo.subdir, cargo.path, 'src', 'lib.rs')):
+        ast.extend(_create_lib(cargo, build))
```

Comment:
This isn't such a big deal, I hope, because at least wrapped crates coming from crates.io have versions and key off of the versions to do a complete meson reconfigure -- because the directory and all its contents have changed.

Maybe for the case of wrap-git we could warn here? Dunno if this is something to reliably detect.

Also, typo: should be reproducibility.

Reply:
I'm not sure how we warn about this here, since we don't (and I don't think should) know whether this is a wrap-git or not at this level

Reply:
Seems very unlikely that a project add or remove its lib, and even less likely to do so without touching Cargo.toml. I guess ninja cannot trigger a reconfigure based on the existence of a file? Or some kind of custom_target that runs meson reconfigure?